Gems/MotionMatching/Code/Source/FeatureMatrixMinMaxScaler.cpp

```diff
@@ -0,0 +1,215 @@
+        result.resize(numRows, numColumns);
+
+        for (FeatureMatrix::Index row = 0; row < numRows; ++row)
+        {
+            for (FeatureMatrix::Index column = 0; column < numColumns; ++column)
+            {
+                result(row, column) = Transform(featureMatrix(row, column), column);
+            }
+        }
+
+        return result;
+    }
+
+    //-------------------------------------------------------------------------
+
+    FeatureMatrix MinMaxScaler::InverseTransform(const FeatureMatrix& featureMatrix) const
+    {
+        const FeatureMatrix::Index numRows = featureMatrix.rows();
+        const FeatureMatrix::Index numColumns = featureMatrix.cols();
+        FeatureMatrix result;
+        result.resize(numRows, numColumns);
+
+        for (FeatureMatrix::Index row = 0; row < numRows; ++row)
+        {
+            for (FeatureMatrix::Index column = 0; column < numColumns; ++column)
+            {
+                result(row, column) = InverseTransform(featureMatrix(row, column), column);
+            }
+        }
+
+        return result;
+    }
+
+    AZ::Vector2 MinMaxScaler::InverseTransform(const AZ::Vector2& value, FeatureMatrix::Index column) const
+    {
+        return AZ::Vector2(
+            InverseTransform(value.GetX(), column + 0),
+            InverseTransform(value.GetY(), column + 1));
+    }
+
+    AZ::Vector3 MinMaxScaler::InverseTransform(const AZ::Vector3& value, FeatureMatrix::Index column) const
+    {
+        return AZ::Vector3(
+            InverseTransform(value.GetX(), column + 0),
+            InverseTransform(value.GetY(), column + 1),
+            InverseTransform(value.GetZ(), column + 2));
+    }
+
+    float MinMaxScaler::InverseTransform(float value, FeatureMatrix::Index column) const
+    {
+        const float normalizedValue = (value - m_featureMin) / m_featureRange;
+        return normalizedValue * m_dataRange[column] + m_dataMin[column];
+    }
+
+    void MinMaxScaler::SaveMinMaxAsCsv(const AZStd::string& filename, const AZStd::vector<AZStd::string>& columnNames)
+    {
+        AZStd::string data;
+
+        // Save column names in the first row
+        if (!columnNames.empty())
+        {
+            for (size_t i = 0; i < columnNames.size(); ++i)
+            {
+                if (i != 0)
+                {
+                    data += ",";
+                }
+
+                data += columnNames[i].c_str();
+            }
+            data += "\n";
+        }
+
+        for (size_t i = 0; i < m_dataMin.size(); ++i)
+        {
+            if (i != 0)
+            {
+                data += ",";
+            }
+
+            data += AZStd::to_string(m_dataMin[i]);
+        }
+
+        data += "\n";
+
+        for (size_t i = 0; i < m_dataMax.size(); ++i)
+        {
+            if (i != 0)
+            {
+                data += ",";
+            }
+
+            data += AZStd::to_string(m_dataMax[i]);
+        }
+
+        data += "\n";
```

Comment:
This CSV is kind of strange.
It is saving the data in rows instead of columns.

Normally it would be something like 

| column name | data min | data max |
|---|---|
| name1 | 0.0 | 10.0 |
| name2 | 1.0 | 15.0 |

Might I interest you in an [AZStd::zip_view](https://en.cppreference.com/w/cpp/ranges/zip_view)

```
for (auto&& [columnName, dataMin, dataMax] : AZStd::ranges::views::zip(columnNames, m_dataMin, m_dataMax)
{
    data += AZStd::string::format("%s,%lf,%lf\n", columnName.c_str(), dataMin, dataMax);
}
```